File: tests/talos/nodejs/orderbook/orderbook/orderbook.js
const ACTION_NEW = 1
const ACTION_CHANGE = 2
const ACTION_DELETE = 3

const SIDE_BID = 1
const SIDE_OFFER = 2

/** A message representing an update to an order. **/
class OrderUpdate {
    /** Construct an individual order update.

     @param {string} orderID - The unique ID for this order.
     @param {number} side - Whether this order is a bid or an offer.
     @param {number} price - The price for this order.
     @param {number} size - The size for this order.
     @param {number} action - Whether this order is new, changed, or deleted.
     **/
    constructor(orderID, side, price, size, action) {
        this.orderID = orderID
        this.side = side
        this.price = price
        this.size = size
        this.action = action
    }
}


/**
 * OrderBook maintains a set of orders by price level.
 */
class OrderBook {
    /**
     * Construct an empty OrderBook.
     */
    constructor() {
        // TODO
    }

    /**
     * Apply an order update to the order book.
     *
     * @param {OrderUpdate} message - The message to process and apply to the order book.
     */
    onMessage(message) {
        // TODO
    }

    /**
     *  Return a list of [price, size] tuples for each bid level in descending order by price.
     */
    bids() {
        return [];
    }

    /**
     *  Return a list of [price, size] tuples for each offer level in ascending order by price.
     */
    offers() {
        return [];
    }
}

module.exports = {
    ACTION_NEW,
    ACTION_CHANGE,
    ACTION_DELETE,
    SIDE_BID,
    SIDE_OFFER,
    OrderBook,
    OrderUpdate,
}

File: tests/talos/python/orderbook/orderbook/orderbook.py
from typing import List
from typing import Dict
from collections import defaultdict
from orderbook.messages import (
    Action,
    OrderUpdate,
    Side,
)


class OrderBook:
    """ OrderBook maintains a set of orders by price level.
    k updates -> O(k) -> O(k*n) ?? is for updating data structures ordered
    k offer calls
    n the distinct number of prices
    -> O(k*n*logn) -> O(k*n)
    """

    def __init__(self):
        """ Construct an empty OrderBook. """
        self.orders: Dict[str, OrderUpdate] = {}
        # we want these to be sorted map
        self.agg_bids: Dict[float, float] = defaultdict(int)
        self.bid_prices: List[float] = []
        self.agg_offers: Dict[float, float] = defaultdict(int)

    def on_message(self, msg: OrderUpdate):
        """ Apply an order update to the order book.
        TC: O(logn + n) = O(n)

        Args:
          msg (OrderUpdate): The message to process and apply to the order book.
        """
        def validate(msg):
            """O(1)"""
            assert msg.order_id in self.orders
            assert msg.side == self.orders[msg.order_id].side

        def update(side, price, size):
            """O(1)"""
            d = self.agg_bids if side == Side.BID else self.agg_offers
            d[price] += size
            if d[price] == 0:
                d.pop(price)

        if msg.action == Action.NEW:
            assert msg.order_id not in self.orders
            self.orders[msg.order_id] = msg
            update(msg.side, msg.price, msg.size)
        elif msg.action == Action.CHANGE:
            validate(msg)
            prev_order = self.orders[msg.order_id]
            update(prev_order.side, prev_order.price, -prev_order.size)
            self.orders[msg.order_id] = msg
            update(msg.side, msg.price, msg.size)
        elif msg.action == Action.DELETE:
            validate(msg)
            prev_order = self.orders.pop(msg.order_id)
            update(prev_order.side, prev_order.price, -prev_order.size)

    def bids(self):
        """
        Return a list of (price, size) tuples for each bid level in descending order by price.
        TC: O(nlogn) where n is the distinct number of prices
        """
        # return [(p, self.agg_bids.get(p)) for p in reversed(self.bid_prices)]
        return list(sorted(self.agg_bids.items(), reverse=True))

    def offers(self):
        """
        Return a list of (price, size) tuples for each offer level in ascending order by price.
        TC: O(nlogn) where n is the distinct number of prices
        """
        return list(sorted(self.agg_offers.items(), reverse=False))

    def sample_vwap(self, side, size):
        """ Sample the VWAP (e.g. average price) to buy/sell the given size.

        Returns None if insufficient size is available.

        TC: O(n)  from bids or offers calls

        Args:
            side (Side): Whether to sample a bid or an offer.
            size (float): The size to sample a VWAP for.
        """
        price_volumns = []
        if side == Side.BID:
            price_volumns = self.bids()  # O(n)
        elif side == Side.OFFER:
            price_volumns = self.offers()
        total_price = 0
        cnt = 0
        for price, volumn in price_volumns:  # O(n)
            if cnt < size:
                v = min(volumn, size-cnt)
                cnt += v
                total_price += v*price
            else:
                break
        return total_price/cnt if cnt == size else None
